talisker/celery.py

```diff
@@ -79,48 +52,35 @@ def signal(sender, **kwargs):
 task_revoked = _counter('revoked')
 
 
-def get_id(body, headers):
-    """get the task id, supporting both celery 4.0.x and 3.1.x"""
-    id = None
-    if 'id' in headers:
-        id = headers['id']
-    elif 'id' in body:
-        id = body['id']
-    return id
-
-
-def before_task_publish(sender, body, headers={}, **kwargs):
-    # TODO: find a way to avoid thread locals
-    if not hasattr(_local, 'timers'):
-        _local.timers = {}
-    name = 'celery.{}.enqueue'.format(sender)
-    timer = talisker.statsd.get_client().timer(name)
-    id = get_id(body, headers)
-    if id is not None:
-        _local.timers[id] = timer
-        timer.start()
+def before_task_publish(sender, body, headers, **kwargs):
+    headers['talisker_enqueue_start'] = time.time()
+    headers['talisker_request_id'] = talisker.request_id.get()
 
 
-def after_task_publish(sender, body, headers={}, **kwargs):
-    id = get_id(body, headers)
-    if id is not None:
-        timer = _local.timers.pop(id)
-        if timer:
-            timer.stop()
+def after_task_publish(sender, body, headers, **kwargs):
+    start_time = headers.pop('talisker_enqueue_start', None)
+    if start_time is not None:
+        ms = (time.time() - start_time) * 1000
+        name = 'celery.{}.enqueue'.format(sender)
+        talisker.statsd.get_client().timing(name, ms)
 
 
 def task_prerun(sender, task_id, task, **kwargs):
     name = 'celery.{}.run'.format(sender.name)
-    task.__talisker_timer = talisker.statsd.get_client().timer(name)
-    task.__talisker_timer.start()
+    task.talisker_timer = talisker.statsd.get_client().timer(name)
+    task.talisker_timer.start()
+    id = task.request.talisker_request_id
+    if id is not None:
+        talisker.request_id.set(id)
 
 
 def task_postrun(sender, task_id, task, **kwargs):
-    task.__talisker_timer.stop()
+    task.talisker_timer.stop()
+    talisker.request_context.clear()
```

Comment:
Hmm, is there a talisker request_context to clear here? Where does the `postrun` signal execute - in the WSGI app or in the celery worker?